heartbeat/storage-mon.in

```diff
@@ -173,77 +205,163 @@ storage-mon_init() {
 	fi
 }
 
-storage-mon_validate() {
-	storage-mon_init
-
-	# Is the state directory writable?
-	state_dir=$(dirname "$OCF_RESKEY_state_file")
-	touch "$state_dir/$$"
-	if [ $? -ne 0 ]; then
-		return $OCF_ERR_CONFIGURED
-	fi
-	rm "$state_dir/$$"
-
-	return $OCF_SUCCESS
-}
-
 storage-mon_monitor() {
-	storage-mon_init
+	if [ -z "$OCF_RESKEY_daemonize" ]; then
+		storage-mon_init
 
-	# Monitor _MUST!_ differentiate correctly between running
-	# (SUCCESS), failed (ERROR) or _cleanly_ stopped (NOT RUNNING).
-	# That is THREE states, not just yes/no.
+		# Monitor _MUST!_ differentiate correctly between running
+		# (SUCCESS), failed (ERROR) or _cleanly_ stopped (NOT RUNNING).
+		# That is THREE states, not just yes/no.
 
-	if [ ! -f "${OCF_RESKEY_state_file}" ]; then
-		return $OCF_NOT_RUNNING
-	fi
+		if [ ! -f "${OCF_RESKEY_state_file}" ]; then
+			return $OCF_NOT_RUNNING
+		fi
 
-	# generate command line
-	cmdline=""
-	for DRIVE in ${OCF_RESKEY_drives}; do
-		cmdline="$cmdline --device $DRIVE --score 1"
-	done
-	cmdline="$cmdline --timeout ${OCF_RESKEY_io_timeout}"
-	if [ -n "${OCF_RESKEY_inject_errors}" ]; then
-		cmdline="$cmdline --inject-errors-percent ${OCF_RESKEY_inject_errors}"
-	fi
-	$STORAGEMON $cmdline
-	if [ $? -ne 0 ]; then
-		status="red"
+		# generate command line
+		cmdline=""
+		for DRIVE in ${OCF_RESKEY_drives}; do
+			cmdline="$cmdline --device $DRIVE --score 1"
+		done
+		cmdline="$cmdline --timeout ${OCF_RESKEY_io_timeout}"
+		if [ -n "${OCF_RESKEY_inject_errors}" ]; then
+			cmdline="$cmdline --inject-errors-percent ${OCF_RESKEY_inject_errors}"
+		fi
+		$STORAGEMON $cmdline
+		if [ $? -ne 0 ]; then
+			status="red"
+		else
+			status="green"
+		fi
+
+		"$ATTRDUP" -n ${ATTRNAME} -U "$status" -d "5s"
```

Comment:
It is good that you added checking for $ATTRDUP initially.
But I guess it is still possible that we fail here somehow - e.g. $ATTRDUP on an unresponsive root so that the actual executable can't be read while directory-structure is already cached.
I guess in general we should error out for anything that can go wrong - e.g. calling binaries - as that we can run the resource with on-failure=fence so that fencing immediately takes us out of a situation where more subtle mechanisms are failing.

Reply:
We're going with score 1 here. So maximum if all disks are bad is number of drives.
Similar as with e.g. $ATTRDUP we could on all exit-codes higher assume that we have some basic issue with the mechanism that evaluates status of drives. That doesn't necessarily mean that the drive is bad. So altering the status to "red" is arguable - maybe leave it at value before as we don't know better - or go red as that would trigger going away from a node that has issues anyway. But on top we might exit monitor with a failure as that we can react on a result that is actually bogus instead of the state of a disk.